lib/charms/opensearch/v0/opensearch_tls.py

```diff
@@ -0,0 +1,270 @@
+from charms.opensearch.v0.opensearch_distro import OpenSearchError
+from charms.opensearch.v0.tls_constants import TLS_RELATION, CertType
+from charms.tls_certificates_interface.v1.tls_certificates import (
+    CertificateAvailableEvent,
+    CertificateExpiringEvent,
+    TLSCertificatesRequiresV1,
+    generate_csr,
+    generate_private_key,
+)
+from ops.charm import ActionEvent, RelationBrokenEvent, RelationJoinedEvent
+from ops.framework import Object
+
+# The unique Charmhub library identifier, never change it
+LIBID = "f4bd9c1dad554f9ea52954b8181cdc19"
+LIBAPI = 0
+LIBPATCH = 0
+
+logger = logging.getLogger(__name__)
+
+
+class OpenSearchTLS(Object):
+    """Class that Manages OpenSearch relation with TLS Certificates Operator."""
+
+    def __init__(self, charm, peer_relation: str):
+        super().__init__(charm, "client-relations")
+
+        self.charm = charm
+        self.peer_relation = peer_relation
+        self.certs = TLSCertificatesRequiresV1(charm, TLS_RELATION)
+
+        self.framework.observe(
+            self.charm.on.set_tls_private_key_action, self._on_set_tls_private_key
+        )
+
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_joined, self._on_tls_relation_joined
+        )
+        self.framework.observe(
+            self.charm.on[TLS_RELATION].relation_broken, self._on_tls_relation_broken
+        )
+
+        self.framework.observe(self.certs.on.certificate_available, self._on_certificate_available)
+        self.framework.observe(self.certs.on.certificate_expiring, self._on_certificate_expiring)
+
+    def _on_set_tls_private_key(self, event: ActionEvent) -> None:
+        """Set the TLS private key, which will be used for requesting the certificate."""
+        cert_type = CertType(event.params["category"])  # type
+        scope = Scope.APP if cert_type == CertType.APP_ADMIN else Scope.UNIT
+
+        if scope == Scope.APP and not self.charm.unit.is_leader():
+            event.log("Only the juju leader unit can set private key for the admin certificates.")
+            return
+
+        try:
+            self._request_certificate(
+                scope, cert_type, event.params.get("key", None), event.params.get("password", None)
+            )
+        except ValueError as e:
+            event.fail(str(e))
+
+    def _on_tls_relation_joined(self, _: RelationJoinedEvent) -> None:
+        """Request certificate when TLS relation joined."""
+        admin_cert = self.charm.secrets.get_object(Scope.APP, CertType.APP_ADMIN)
+        if self.charm.unit.is_leader() and admin_cert is None:
+            self._request_certificate(Scope.APP, CertType.APP_ADMIN)
+
+        self._request_certificate(Scope.UNIT, CertType.UNIT_TRANSPORT)
+        self._request_certificate(Scope.UNIT, CertType.UNIT_HTTP)
+
+    def _on_tls_relation_broken(self, _: RelationBrokenEvent) -> None:
+        """We don't need to handle much here."""
```

Comment:
Thanks Will - I missed handling a case serving as a safeguard here: if all certificates (3 per unit) have already been provided -> nothing to handle, we keep the certificates stored to guarantee the continuity of functioning of the node until they hit 1 week before their expiration date. 
BUT - in case only a subset of those requested certs is available, it means that TLS will not work - thus requiring the unit to be put on a BlockedStatus